nidx/nidx_relation/src/graph_query_parser.rs

```diff
@@ -425,68 +456,95 @@ impl From<&str> for Term {
+                        Some(source_pb) => Node::try_from(source_pb)?,
+                        None => Node::default(),
+                    };
+                    let relation = match &path.relation {
+                        Some(relation_pb) => Relation::try_from(relation_pb)?,
+                        None => Relation::default(),
+                    };
+                    let destination = match &path.destination {
+                        Some(destination_pb) => Node::try_from(destination_pb)?,
+                        None => Node::default(),
+                    };
+
+                    let path_query = if path.undirected {
+                        GraphQuery::PathQuery(PathQuery::UndirectedPath((
+                            Expression::Value(source),
+                            Expression::Value(relation),
+                            Expression::Value(destination),
+                        )))
+                    } else {
+                        GraphQuery::PathQuery(PathQuery::DirectedPath((
+                            Expression::Value(source),
+                            Expression::Value(relation),
+                            Expression::Value(destination),
+                        )))
+                    };
+
+                    BooleanExpression::Literal(path_query)
+                }
 
-            nidx_protos::graph_query::Query::Relation(relation) => {
-                Self::RelationQuery(RelationQuery(Expression::Value(Relation::try_from(relation)?)))
-            }
+                nidx_protos::graph_query::path_query::Query::BoolNot(bool_not) => {
+                    let subquery = BoolGraphQuery::try_from(bool_not.as_ref())?.0;
+                    BooleanExpression::Not(Box::new(subquery))
+                }
 
-            nidx_protos::graph_query::Query::Path(path) => {
-                let source = match &path.source {
-                    Some(source_pb) => Node::try_from(source_pb)?,
-                    None => Node::default(),
-                };
-                let relation = match &path.relation {
-                    Some(relation_pb) => Relation::try_from(relation_pb)?,
-                    None => Relation::default(),
-                };
-                let destination = match &path.destination {
-                    Some(destination_pb) => Node::try_from(destination_pb)?,
-                    None => Node::default(),
-                };
+                nidx_protos::graph_query::path_query::Query::BoolAnd(bool_and) => {
+                    let mut operands = Vec::with_capacity(bool_and.operands.len());
+                    for operand in bool_and.operands.iter() {
+                        let subquery = BoolGraphQuery::try_from(operand)?.0;
+                        operands.push(subquery);
+                    }
+                    BooleanExpression::Operation(BooleanOperation {
+                        operator: Operator::And,
+                        operands,
+                    })
```

Comment:
Have you tried doing this with an iterator?

```suggestion
                    BooleanExpression::Operation(BooleanOperation {
                        operator: Operator::And,
                        operands: bool_and
                            .operands
                            .iter()
                            .map(|o| BoolGraphQuery::try_from(o).map(|x| x.0))
                            .collect::<anyhow::Result<Vec<_>>>()?,
                    })
```